lib/services/access_request_cache_test.go

```diff
@@ -0,0 +1,245 @@
+				"00000000-0000-0000-0000-000000000003",
+				"00000000-0000-0000-0000-000000000004",
+				"00000000-0000-0000-0000-000000000005",
+			},
+		},
+		{
+			Sort:       proto.AccessRequestSort_STATE,
+			Descending: false,
+			Expect: []string{
+				"00000000-0000-0000-0000-000000000002", // approved
+				"00000000-0000-0000-0000-000000000004", // approved
+				"00000000-0000-0000-0000-000000000003", // denied
+				"00000000-0000-0000-0000-000000000001", // pending
+				"00000000-0000-0000-0000-000000000005", // pending
+			},
+		},
+		{
+			Sort:       proto.AccessRequestSort_STATE,
+			Descending: true,
+			Expect: []string{
+				"00000000-0000-0000-0000-000000000005", // pending
+				"00000000-0000-0000-0000-000000000001", // pending
+				"00000000-0000-0000-0000-000000000003", // denied
+				"00000000-0000-0000-0000-000000000004", // approved
+				"00000000-0000-0000-0000-000000000002", // approved
+			},
+		},
+	}
+
+	created := time.Now()
+	for _, rr := range rrs {
+		r, err := types.NewAccessRequest(rr.id, "alice@example.com", "some-role")
+		require.NoError(t, err)
+		require.NoError(t, r.SetState(rr.state))
+		r.SetCreationTime(created.UTC())
+		created = created.Add(time.Second)
+		_, err = svcs.CreateAccessRequestV2(ctx, r)
+		require.NoError(t, err)
+	}
+
+	timeout := time.After(time.Second * 30)
+
+	for {
+		rsp, err := cache.ListAccessRequests(ctx, &proto.ListAccessRequestsRequest{
+			Limit: int32(len(rrs)),
+		})
+		require.NoError(t, err)
+		if len(rsp.AccessRequests) == len(rrs) {
+			break
+		}
+
+		select {
+		case <-timeout:
+			require.FailNow(t, "timeout waiting for access request cache to populate")
+		case <-time.After(time.Millisecond * 200):
+		}
+	}
+
+	for _, tt := range tts {
+
+		var nextKey string
+		var reqIDs []string
+		for {
+			rsp, err := cache.ListAccessRequests(ctx, &proto.ListAccessRequestsRequest{
+				StartKey:   nextKey,
+				Limit:      3,
+				Sort:       tt.Sort,
+				Descending: tt.Descending,
+			})
+			require.NoError(t, err)
+			for _, r := range rsp.AccessRequests {
+				reqIDs = append(reqIDs, r.GetName())
+			}
+			nextKey = rsp.NextKey
+			if nextKey == "" {
+				break
+			}
+		}
+
+		require.Equal(t, tt.Expect, reqIDs, "index=%s, descending=%v", tt.Sort.String(), tt.Descending)
+	}
+
+	// verify that delete events are correctly processed
+	timeout = time.After(time.Second * 30)
+	for i, rr := range rrs {
+		require.NoError(t, svcs.DeleteAccessRequest(ctx, rr.id))
+	WaitForReplication:
+		for {
+			rsp, err := cache.ListAccessRequests(ctx, &proto.ListAccessRequestsRequest{
+				Limit: int32(len(rrs)),
+			})
+			require.NoError(t, err)
+			if len(rsp.AccessRequests) == len(rrs)-(i+1) {
+				break WaitForReplication
+			}
+
+			select {
+			case <-timeout:
+				require.FailNow(t, "timeout waiting for cache to to reach expected resource count", "have=%d, want=%d", len(rsp.AccessRequests), len(rrs)-(i+1))
+			case <-time.After(time.Millisecond * 200):
+			}
+		}
```

Comment:
It can be, but I kind of hate `require.EventuallyT`.  Assertions are interpreted as condition failures rather than test failures in there, and if you actually do end up causing an explicit failure it actually just panics.  IMO it's inferior to just using a normal loop.  If you have a strong preference I will change, but otherwise I'm going to leave as-is.

Reply:
no strong preference here